numba/targets/arrayobj.py

```diff
@@ -1064,8 +1064,43 @@ def _attempt_nocopy_reshape(context, builder, aryty, ary, newnd, newshape,
                             ary.itemsize, is_f_order])
     return res
 
+
+def normalize_reshape_value(origsize, shape):
+    num_neg_value = 0
+    for s in shape:
+        if s < 0:
+            num_neg_value += 1
+
+    if num_neg_value == 0:
+        total_size = 1
+        for s in shape:
+            total_size *= s
+
+        if origsize != total_size:
+            raise ValueError("total size of new array must be unchanged")
+
+    elif num_neg_value == 1:
+        # infer negative dimension
+        known_size = 1
+        inferred_ax = 0
+        for ax, s in enumerate(shape):
+            if s > 0:
+                known_size *= s
+            else:
+                inferred_ax = ax
+
+        if origsize % known_size > 0:
+            raise ValueError("total size of new array must be unchanged")
```

Comment:
OK, I've confirmed that this is the same error that NumPy raises, so this is fine.